posthog/temporal/workflows/snowflake_batch_export.py

```diff
@@ -65,23 +101,142 @@ class SnowflakeInsertInputs:
+        warehouse=inputs.warehouse,
+        database=inputs.database,
+        schema=inputs.schema,
+        role=inputs.role,
+    ) as connection:
+        use_namespace(connection, inputs.database, inputs.schema)
+        connection.cursor().execute("SET ABORT_DETACHED_QUERY = FALSE")
+
+        yield connection
+
+
+async def execute_async_query(
+    connection: SnowflakeConnection,
+    query: str,
+    parameters: dict | None = None,
+    file_stream=None,
+    poll_interval: float = 1.0,
+) -> str:
+    """Wrap Snowflake connector's polling API in a coroutine.
+
+    This enables asynchronous execution of queries to release the event loop to execute other tasks
+    while we poll for a query to be done. For example, the event loop may use this time for heartbeating.
+
+    Args:
+        connection: A SnowflakeConnection object as produced by snowflake.connector.connect.
+        query: A query string to run asynchronously.
+        parameters: An optional dictionary of parameters to bind to the query.
+        poll_interval: Specify how long to wait in between polls.
+    """
+    cursor = connection.cursor()
+
+    # Snowflake docs incorrectly state that the 'params' argument is named 'parameters'.
+    result = cursor.execute_async(query, params=parameters, file_stream=file_stream)
+    query_id = cursor.sfqid or result["queryId"]
+    query_status = None
+
+    try:
+        query_status = connection.get_query_status_throw_if_error(query_id)
+
+        while connection.is_still_running(query_status):
+            query_status = connection.get_query_status_throw_if_error(query_id)
+            await asyncio.sleep(poll_interval)
+
+    except snowflake.connector.ProgrammingError:
+        # TODO: logging? Other handling?
+        raise
+
+    return query_id
+
+
+async def create_table_in_snowflake(connection: SnowflakeConnection, table_name: str) -> None:
+    """Asynchronously create the table if it doesn't exist.
+
+    Note that we use the same schema as the snowflake-plugin for backwards compatibility."""
+    await execute_async_query(
+        connection,
+        f"""
+        CREATE TABLE IF NOT EXISTS "{table_name}" (
+            "uuid" STRING,
+            "event" STRING,
+            "properties" VARIANT,
+            "elements" VARIANT,
+            "people_set" VARIANT,
+            "people_set_once" VARIANT,
+            "distinct_id" STRING,
+            "team_id" INTEGER,
+            "ip" STRING,
+            "site_url" STRING,
+            "timestamp" TIMESTAMP
+        )
+        COMMENT = 'PostHog generated events table'
+        """,
+    )
+
+
+async def put_file_to_snowflake_table(
+    connection: SnowflakeConnection,
+    file: BatchExportTemporaryFile,
+    table_name: str,
+    file_no: int,
+):
     """Executes a PUT query using the provided cursor to the provided table_name.
 
+    Sadly, Snowflake's execute_async does not work with PUT statements. So, we pass the execute
```

Comment:
Totally aside, it's wild to me that their execute fn does or doesn't work only with certain statements. Weird!

Reply:
There is some explanation of why here: https://github.com/snowflakedb/snowflake-connector-python/issues/1227#issuecomment-1281479236.